pkg/dns/ibm/private/dnssvcs_provider.go

```diff
@@ -0,0 +1,273 @@
+		return fmt.Errorf("delete: invalid result")
+	}
+	for _, target := range record.Spec.Targets {
+		for _, resourceRecord := range result.ResourceRecords {
+
+			var resourceRecordTarget string
+			rData, ok := resourceRecord.Rdata.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("delete: failed to get resource data: %v", resourceRecord.Rdata)
+			}
+			if *resourceRecord.Type == string(iov1.CNAMERecordType) {
+				resourceRecordTarget = fmt.Sprint(rData["cname"])
+			}
+			if *resourceRecord.Type == string(iov1.ARecordType) {
+				resourceRecordTarget = fmt.Sprint(rData["ip"])
+			}
+
+			if resourceRecordTarget == target && *resourceRecord.Name == DNSName {
+				delOpt := p.dnsService.NewDeleteResourceRecordOptions(p.config.InstanceID, zone.ID, *resourceRecord.ID)
+				delResponse, err := p.dnsService.DeleteResourceRecord(delOpt)
+				if err != nil {
+					if delResponse != nil && delResponse.StatusCode != http.StatusNotFound {
+						return fmt.Errorf("delete: failed to delete the dns record: %w", err)
+					}
+				}
+				if delResponse != nil && delResponse.StatusCode != http.StatusNotFound {
+					log.Info("deleted DNS record", "record", record, "zone", zone, "target", target)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateDNSServices validates that provider clients can communicate with
+// associated API endpoints by having each client list zones of the instance.
+func validateDNSServices(provider *Provider) error {
+	var errs []error
+
+	for _, zoneID := range provider.config.Zones {
+		getDnszoneOptions := provider.dnsService.NewGetDnszoneOptions(
+			provider.config.InstanceID,
+			zoneID)
+		_, _, err := provider.dnsService.GetDnszone(getDnszoneOptions)
+		if err != nil {
+			return fmt.Errorf("validateDNSServices: failed to get dns zone: %w", err)
+		}
+
+		listOpt := provider.dnsService.NewListResourceRecordsOptions(provider.config.InstanceID, zoneID)
+		listResult, response, err := provider.dnsService.ListResourceRecords(listOpt)
+		if err != nil {
+			if response != nil && response.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("validateDNSServices: failed to list dns records: %w", err)
+			}
+		}
+		if listResult == nil {
+			return fmt.Errorf("validateDNSServices: invalid result")
+		}
+
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	if err := validateInputDNSData(record, zone); err != nil {
+		return fmt.Errorf("createOrUpdateDNSRecord: invalid dns input data: %w", err)
+	}
+
+	listOpt := p.dnsService.NewListResourceRecordsOptions(p.config.InstanceID, zone.ID)
+	// Some dns records (e.g. wildcard record) have an ending "." character in the DNSName
+	DNSName := strings.TrimSuffix(record.Spec.DNSName, ".")
+
+	// TTL must be one of [1 60 120 300 600 900 1800 3600 7200 18000 43200]
+	useDefaultTTL := true
+	for _, v := range []int64{1, 60, 120, 300, 600, 900, 1800, 3600, 7200, 18000, 43200} {
```

Comment:
These TTL constraints come from the actual dns services. Unfortunately they are inconsistent.

In case of CIS, TTL must be between 120 and 2,147,483,647 seconds, or 1 for Automatic. The code makes sure that in case record.Spec.RecordTTL has an invalid value (0 or 2-119) then the default TTL value will be used.